components/ai_chat/renderer/page_content_extractor.cc

```diff
@@ -0,0 +1,223 @@
+#include "base/containers/span.h"
+#include "base/functional/bind.h"
+#include "base/strings/string_piece_forward.h"
+#include "base/values.h"
+#include "brave/components/ai_chat/common/page_content_extractor.mojom-shared.h"
+#include "brave/components/ai_chat/common/page_content_extractor.mojom.h"
+#include "brave/components/ai_chat/renderer/page_text_distilling.h"
+#include "content/public/renderer/render_frame.h"
+#include "content/public/renderer/render_frame_observer.h"
+#include "net/base/url_util.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+#include "third_party/blink/public/platform/web_string.h"
+#include "third_party/blink/public/web/web_local_frame.h"
+#include "third_party/blink/public/web/web_script_source.h"
+#include "url/gurl.h"
+#include "url/url_constants.h"
+#include "v8/include/v8-isolate.h"
+
+namespace {
+
+const char16_t kYoutubeTranscriptUrlExtractionScript[] =
+    uR"JS(
+      (function() {
+        return ytplayer?.config?.args?.raw_player_response?.captions?.playerCaptionsTracklistRenderer?.captionTracks?.[0]?.baseUrl
+      })()
+    )JS";
+
+const char16_t kVideoTrackTranscriptUrlExtractionScript[] =
+    // TODO(petemill): Make more informed srclang choice.
+    // TODO(petemill): Observe <video>.textTracks
+    uR"JS(
+      (function() {
+        const nodes = document.querySelectorAll('video track')
+        if (nodes.length) {
+          let selectedNode = nodes[0]
+          for (const node of nodes) {
+            if (node.srclang.toLowerCase() === 'en') {
+              selectedNode = node
+            }
+          }
+          return node.src
+        }
+      })()
+    )JS";
+
+constexpr auto kYouTubeHosts = base::MakeFixedFlatSet<base::StringPiece>(
+    {"youtube.com", "www.youtube.com", "m.youtube.com"});
+
+// TODO(petemill): Use heuristics to determine if page's main focus is
+// a video, and not a hard-coded list of Url hosts.
+constexpr auto kVideoTrackHosts =
+    base::MakeFixedFlatSet<base::StringPiece>({"www.ted.com"});
+
+}  // namespace
+
+namespace ai_chat {
+
+PageContentExtractor::PageContentExtractor(
+    content::RenderFrame* render_frame,
+    service_manager::BinderRegistry* registry,
+    int32_t global_world_id,
+    int32_t isolated_world_id)
+    : content::RenderFrameObserver(render_frame),
+      global_world_id_(global_world_id),
+      isolated_world_id_(isolated_world_id),
+      weak_ptr_factory_(this) {
+  if (!render_frame->IsMainFrame()) {
+    return;
+  }
+  // Bind mojom API to allow browser to communicate with this class
+  // Being a RenderFrameObserver, this object is scoped to the RenderFrame.
+  // Unretained is safe here because `registry` is also scoped to the
+  // RenderFrame.
+  registry->AddInterface(base::BindRepeating(
+      &PageContentExtractor::BindReceiver, base::Unretained(this)));
+}
+
+PageContentExtractor::~PageContentExtractor() = default;
+
+void PageContentExtractor::OnDestruct() {
+  delete this;
```

Comment:
Yes, especially in RenderFrameObservers. See all the overrides here https://source.chromium.org/chromium/chromium/src/+/main:content/public/renderer/render_frame_observer.h;l=78;bpv=1;bpt=1?q=RenderFrameObserver
But you also see this in other parts of Chromium code. My guess is that when it's obvious that there won't be other threads accessing the object, it's ok to use `delete`. Otherwise DeleteSoon can be used if you've passed pointers elsewhere. Here, we don't pass pointers and we must ensure delete happens when the RenderFrame is itself deleted.

Reply:
Well, I don't think it's just about other threads. Something on the same thread in the call stack might have a raw pointer.  It looks like `PageContentExtractor` isn't owned by a resource? and therefore it's self-owned. 

I'm not sure what the protocol here is but consistency with the existing code matters too. I'm seeing some overrides with `delete this;` and some with `base::SingleThreadTaskRunner`

Reply:
Chromium's and Brave's RenderFrameObservers call `delete this` in `OnDestruct`, so I don't think there're consistency issues being introduced here. No references are being passed from this instance to anything that isn't owned by this instance or not provided via a WeakPtr. If I had to guess the reason `delete` is used, it's because it is important not to access `render_frame()` after `OnDestruct` has happened.